lock timelocked utxos on maker startup

undeath opened the pull request:
A different fix for #955 because the issue cannot be fixed by #957 (which still adds an unused feature now) as explained by @AdamISZ: https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/957#issuecomment-892486730

~~This PR is WIP because the newly added test is failing. Something with the FidelityBonds code is acting weirdly and I have no idea how/what to fix. @chris-belcher any idea what's wrong here?~~

```
    def add_utxo(self, txid, index, script, value, height=None):
        assert isinstance(txid, bytes)
        assert isinstance(index, Integral)
        assert isinstance(script, bytes)
        assert isinstance(value, Integral)
    
        if script not in self._script_map:
>           raise WalletError("Tried to add UTXO for unknown key to wallet.")
E           jmclient.wallet.WalletError: Tried to add UTXO for unknown key to wallet.
```

Note the commits "clean up timestamp_to_time_number()" and "fixup 6cc1893" are cherry-picked from #957

chris-belcher commented:
Doesn't this already happen? I'm pretty sure the wallet code already adds timelocked UTXOs to the frozen list.

AdamISZ commented:
> Doesn't this already happen? I'm pretty sure the wallet code already adds timelocked UTXOs to the frozen list.

Yes but the issue is that if the timelocked utxo has passed its timelock, it doesn't get frozen.

chris-belcher commented:
But it will *remain* frozen from the time it was still locked(?) Unless the user unfreezes it

AdamISZ commented:
> But it will remain frozen from the time it was still locked(?) Unless the user unfreezes it

Two answers here: we may have intended that, but I believe the code as-is always checks the timelock, then unfreezes if it's past it, in syncing.

Second, if we removed that auto-unfreeze action, it would still be needed, imo, to address the case where a user manually unfroze it and then started the maker, thinking it would work, but it will actually either crash or result in failed joins silently (I think by default the latter, but neither are OK).

AdamISZ commented:
OK, on checking more carefully, the situation is:

`FidelityBondMixin.add_utxo`, called by `WalletService._add_unspent_txo` as part of `WalleService.sync_unspent` in the case of a fidelity bond enabled wallet, will freeze the utxo (call `BaseWallet.disable_utxo`, which, note, persists this change to the wallet file) *if and only if* the current time is later than the timelock (ie coin is still timelocked). Otherwise, nothing is done. The only other place that `BaseWallet.disable_utxo` is called is via the manual `freeze` method in `wallet_utils.py`.

So currently , as the timelock passes:

* If the user does nothing, and has in the past opened the wallet in non-read-only mode, the utxo will remain frozen.
* If at some time after the timelock, the user manually unfreezes the now-time-unlocked coin, it should stay unfrozen from then on.

What this PR about (at least as I understood) is changing that second bullet point: *if* a maker is started, we should in fact auto-refreeze the now-time-unlocked coin for the duration of the maker's running to avoid either a crash, or silent failure (likely repeated) to complete coinjoins.

AdamISZ commented on jmclient/jmclient/maker.py:
```diff
@@ -254,6 +255,23 @@ def modify_orders(self, to_cancel, to_announce):
+        Freeze all wallet's timelocked UTXOs. These cannot be spent in a
+        coinjoin because of protocol limitations.
+        """
+        if not hasattr(self.wallet_service.wallet, 'FIDELITY_BOND_MIXDEPTH'):
```
is `isinstance` not viable here? I thought that worked OK for mixins?

undeath replied:
`isinstance` is not pythonic. It violates the duck-typing philosophy ("If it quacks like a duck it is a duck."). It makes sense for base types (that don't have identifying attributes) but usually not for more complex objects.

AdamISZ commented:
Apart from my nit above, this looks to address the issue, so at least conceptACK for now until we sort the tests.

chris-belcher commented:
Thanks for explaining. Concept ACK.

Reading the discussion on IRC about this, I suggest just printing out a warning to tell the user that the UTXO was frozen again.

AdamISZ commented:
@undeath is there not a case for doing a wallet change on exit? Like, something similar to the current `atexit.register` in the `Storage` class (even go as far as having a hook into that shutdown processing).

As we discussed on IRC, while it's true that in a crash scenario it might stay frozen, there's surely still a case for having a maker-user not be forced to re-unfreeze a coin he's already un-frozen, in anticipation of spending it.

AdamISZ commented on jmclient/test/test_maker.py:
```diff
@@ -164,6 +166,23 @@ def test_verify_unsigned_tx_nonsw_valid(setup_env_nodeps):
+    wallet = SegwitWalletFidelityBonds(storage)
+    ts = wallet.datetime_to_time_number(
+        datetime.datetime.strptime("2021-07", "%Y-%m"))
+    tl_path = wallet.get_path(0, wallet.BIP32_TIMELOCK_ID, 0, ts)
```
Index 18, not index 0, for that month.

AdamISZ commented on jmclient/jmclient/maker.py:
```diff
@@ -254,6 +255,23 @@ def modify_orders(self, to_cancel, to_announce):
+            return
+
+        md_utxos = self.wallet_service.get_utxos_by_mixdepth()
+        for tx, details in md_utxos[self.wallet_service.FIDELITY_BOND_MIXDEPTH]:
```
missing `.items()` to take key/value

AdamISZ commented on jmclient/jmclient/maker.py:
```diff
@@ -254,6 +255,23 @@ def modify_orders(self, to_cancel, to_announce):
+        md_utxos = self.wallet_service.get_utxos_by_mixdepth()
+        for tx, details in md_utxos[self.wallet_service.FIDELITY_BOND_MIXDEPTH]:
+            if self.wallet_service.is_timelocked_path(details['path'])\
+                    and not self.wallet_service.is_disabled(details['path']):
```
You can fix this with `self.wallet_service._utxos.is_disabled(*tx)`, or something similar.

AdamISZ commented:
OK it looks like you fixed at literally the same time as my comment :) But the actual code tested still needs those other two fixes I think, for it to pass.

undeath commented:
Just pushed an update to fix the issues you just found @AdamISZ 

Also added an `atexit` handler to unfreeze on shutdown.

Additionally I cleaned up the `timenumber` argument in the wallet. Everywhere in the code it was set to `index = timenumber` (except for one test case), thus the argument was not needed and only added confusion. I guess it was a leftover from when the timelocked addresses were generated using indices and not purely based on the time itself.

AdamISZ commented:
Confirmed that the test passes with the changes, but:

* get an error from the `at_exit`, you are calling `enable_utxo` method of BaseWallet but it doesn't exist, you can call `toggle_disable_utxo` or add enable_utxo at that level or call it on the utxomanager object.

On the removal of the extra timenumber arg, I remarked on that to @chris-belcher at the time of merge but he didn't change it (I saw it as a minor nit). Not sure we should do that here although I guess I don't mind the extra review.

undeath commented:
ah, confused the UTXOManager methods with the wallet methods. Should be fixed now.

AdamISZ commented:
OK I'm confident that the timenumber change is OK.
Will do some regtest testing now.

AdamISZ commented:
Successful manual tests on regtest (TLU = timelock-type utxos):

:heavy_check_mark: TLU pre-TL are frozen as expected (just double-checking pre-existing behaviour)
:heavy_check_mark: TLU post-TL, checked that their frozen-ness state persists across calls to `wallet_tool.py freeze`, as expected
:heavy_check_mark: with only 1 TLU post-TL, checked that as expected, maker refuses to start based on "no coins"
:heavy_check_mark: with 1 TLU post-TL and one vanilla UTXO, checked that maker starts and prints the warning
:heavy_check_mark: with wallet as last point: yield generator starts successfully and prints offer based on non-TLU coin(s)
:heavy_check_mark: with same wallet: try sendpayment from taker with amount < value of non-TLU was successful.
:heavy_check_mark: checked that behaviour of wallet is as expected with respect to spending and freezing, in Joinmarket-Qt

tACK

undeath commented:
squashed

undeath commented:
rebased
